# MSDS 7337 - Natural Language Processing

## Homework 2

### by Evangelos GIakoumakis

Import required libraries to perform required tasks.

In [1]:
from __future__ import division  # Python 2 users only
import nltk, re, pprint
from nltk import word_tokenize
import urllib

Print out system information.

In [5]:
print ("System information:")
import platform; 
print(platform.platform())
import sys
print("Python", sys.version)
max_range = 101

System information:
Windows-10-10.0.17134
('Python', '2.7.15 |Anaconda, Inc.| (default, May  1 2018, 18:37:09) [MSC v.1500 64 bit (AMD64)]')


Function to download texts given a URL address and place in a string.

In [6]:
def text_fetcher(url):

    response = urllib.urlopen(url)
    raw = response.read().decode('utf8')
    if len(raw) > 0:
        return raw
    else:
        return "Failed"

List of URL's containing the address of 100 books in txt format from the gutenberg website.

In [7]:
url_list = list(range(0, max_range)) 

url_list[0] = "http://www.gutenberg.org/cache/epub/7841/pg7841.txt"
url_list[1] = "http://www.gutenberg.org/cache/epub/5742/pg5742.txt"
url_list[2] = "http://www.gutenberg.org/cache/epub/13539/pg13539.txt"
url_list[3] = "http://www.gutenberg.org/cache/epub/7425/pg7425.txt"
url_list[4] = "http://www.gutenberg.org/cache/epub/16046/pg16046.txt"
url_list[5] = "http://www.gutenberg.org/cache/epub/22420/pg22420.txt"
url_list[6] = "http://www.gutenberg.org/files/23424/23424-0.txt"
url_list[7] = "http://www.gutenberg.org/cache/epub/18217/pg18217.txt"
url_list[8] = "http://www.gutenberg.org/cache/epub/24053/pg24053.txt"
url_list[9] = "http://www.gutenberg.org/cache/epub/24644/pg24644.txt"
url_list[10] = "http://www.gutenberg.org/cache/epub/24703/pg24703.txt"
url_list[11] = "http://www.gutenberg.org/cache/epub/15659/pg15659.txt"
url_list[12] = "http://www.gutenberg.org/cache/epub/15170/pg15170.txt"
url_list[13] = "http://www.gutenberg.org/cache/epub/10811/pg10811.txt"
url_list[14] = "http://www.gutenberg.org/cache/epub/9106/pg9106.txt"
url_list[15] = "http://www.gutenberg.org/cache/epub/19721/pg19721.txt"
url_list[16] = "http://www.gutenberg.org/cache/epub/14640/pg14640.txt"
url_list[17] = "http://www.gutenberg.org/cache/epub/14668/pg14668.txt"
url_list[18] = "http://www.gutenberg.org/cache/epub/14766/pg14766.txt"
url_list[19] = "http://www.gutenberg.org/cache/epub/14880/pg14880.txt"
url_list[20] = "http://www.gutenberg.org/cache/epub/15040/pg15040.txt"
url_list[21] = "http://www.gutenberg.org/cache/epub/16751/pg16751.txt"
url_list[22] = "http://www.gutenberg.org/files/13853/13853-0.txt"
url_list[23] = "http://www.gutenberg.org/cache/epub/15825/pg15825.txt"
url_list[24] = "http://www.gutenberg.org/cache/epub/18561/pg18561.txt"
url_list[25] = "http://www.gutenberg.org/cache/epub/18702/pg18702.txt"
url_list[26] = "http://www.gutenberg.org/cache/epub/19923/pg19923.txt"
url_list[27] = "http://www.gutenberg.org/files/22795/22795-0.txt"
url_list[28] = "http://www.gutenberg.org/cache/epub/16936/pg16936.txt"
url_list[29] = "http://www.gutenberg.org/cache/epub/9078/pg9078.txt"
url_list[30] = "http://www.gutenberg.org/cache/epub/6685/pg6685.txt"
url_list[31] = "http://www.gutenberg.org/cache/epub/18909/pg18909.txt"
url_list[32] = "http://www.gutenberg.org/cache/epub/19469/pg19469.txt"
url_list[33] = "http://www.gutenberg.org/cache/epub/11023/pg11023.txt"
url_list[34] = "http://www.gutenberg.org/cache/epub/15353/pg15353.txt"
url_list[35] = "http://www.gutenberg.org/cache/epub/22065/pg22065.txt"
url_list[36] = "http://www.gutenberg.org/cache/epub/24072/pg24072.txt"
url_list[37] = "http://www.gutenberg.org/cache/epub/15626/pg15626.txt"
url_list[38] = "http://www.gutenberg.org/files/23424/23424-0.txt"
url_list[39] = "http://www.gutenberg.org/cache/epub/15577/pg15577.txt"
url_list[40] = "http://www.gutenberg.org/files/18274/18274-0.txt"
url_list[41] = "http://www.gutenberg.org/cache/epub/19381/pg19381.txt"
url_list[42] = "http://www.gutenberg.org/cache/epub/13347/pg13347.txt"
url_list[43] = "http://www.gutenberg.org/files/21973/21973-0.txt"
url_list[44] = "http://www.gutenberg.org/cache/epub/10737/pg10737.txt"
url_list[45] = "http://www.gutenberg.org/cache/epub/21266/pg21266.txt"
url_list[46] = "http://www.gutenberg.org/cache/epub/17365/pg17365.txt"
url_list[47] = "http://www.gutenberg.org/cache/epub/2284/pg2284.txt"
url_list[48] = "http://www.gutenberg.org/cache/epub/3031/pg3031.txt"
url_list[49] = "http://www.gutenberg.org/cache/epub/23667/pg23667.txt"
url_list[50] = "http://www.gutenberg.org/files/15950/15950-0.txt"
url_list[51] = "http://www.gutenberg.org/files/2441/2441-0.txt"
url_list[52] = "http://www.gutenberg.org/cache/epub/17185/pg17185.txt"
url_list[53] = "http://www.gutenberg.org/cache/epub/10834/pg10834.txt"
url_list[54] = "http://www.gutenberg.org/cache/epub/18790/pg18790.txt"
url_list[55] = "http://www.gutenberg.org/cache/epub/16140/pg16140.txt"
url_list[56] = "http://www.gutenberg.org/cache/epub/21948/pg21948.txt"
url_list[57] = "http://www.gutenberg.org/cache/epub/10726/pg10726.txt"
url_list[58] = "http://www.gutenberg.org/cache/epub/18767/pg18767.txt"
url_list[59] = "http://www.gutenberg.org/cache/epub/22408/pg22408.txt"
url_list[60] = "http://www.gutenberg.org/cache/epub/14759/pg14759.txt"
url_list[61] = "http://www.gutenberg.org/cache/epub/23576/pg23576.txt"
url_list[62] = "http://www.gutenberg.org/cache/epub/24263/pg24263.txt"
url_list[63] = "http://www.gutenberg.org/files/24222/24222-0.txt"
url_list[64] = "http://www.gutenberg.org/cache/epub/23941/pg23941.txt"
url_list[65] = "http://www.gutenberg.org/cache/epub/18525/pg18525.txt"
url_list[66] = "http://www.gutenberg.org/cache/epub/24409/pg24409.txt"
url_list[67] = "http://www.gutenberg.org/cache/epub/24852/pg24852.txt"
url_list[68] = "http://www.gutenberg.org/cache/epub/24598/pg24598.txt"
url_list[69] = "http://www.gutenberg.org/cache/epub/11313/pg11313.txt"
url_list[70] = "http://www.gutenberg.org/cache/epub/18561/pg18561.txt"
url_list[71] = "http://www.gutenberg.org/cache/epub/16667/pg16667.txt"
url_list[72] = "http://www.gutenberg.org/cache/epub/20107/pg20107.txt"
url_list[73] = "http://www.gutenberg.org/cache/epub/22727/pg22727.txt"
url_list[74] = "http://www.gutenberg.org/cache/epub/9995/pg9995.txt"
url_list[75] = "http://www.gutenberg.org/cache/epub/22911/pg22911.txt"
url_list[76] = "http://www.gutenberg.org/cache/epub/24676/pg24676.txt"
url_list[77] = "http://www.gutenberg.org/cache/epub/24884/pg24884.txt"
url_list[78] = "http://www.gutenberg.org/files/21783/21783-0.txt"
url_list[79] = "http://www.gutenberg.org/files/17160/17160-0.txt"
url_list[80] = "http://www.gutenberg.org/cache/epub/16728/pg16728.txt"
url_list[81] = "http://www.gutenberg.org/cache/epub/16379/pg16379.txt"
url_list[82] = "http://www.gutenberg.org/cache/epub/20522/pg20522.txt"
url_list[83] = "http://www.gutenberg.org/cache/epub/20698/pg20698.txt"
url_list[84] = "http://www.gutenberg.org/cache/epub/19423/pg19423.txt"
url_list[85] = "http://www.gutenberg.org/cache/epub/7803/pg7803.txt"
url_list[86] = "http://www.gutenberg.org/cache/epub/24858/pg24858.txt"
url_list[87] = "http://www.gutenberg.org/cache/epub/22245/pg22245.txt"
url_list[88] = "http://www.gutenberg.org/cache/epub/23560/pg23560.txt"
url_list[89] = "http://www.gutenberg.org/cache/epub/23728/pg23728.txt"
url_list[90] = "http://www.gutenberg.org/cache/epub/22425/pg22425.txt"
url_list[91] = "http://www.gutenberg.org/cache/epub/22600/pg22600.txt"
url_list[92] = "http://www.gutenberg.org/cache/epub/22766/pg22766.txt"
url_list[93] = "http://www.gutenberg.org/files/22917/22917-0.txt"
url_list[94] = "http://www.gutenberg.org/cache/epub/12655/pg12655.txt"
url_list[95] = "http://www.gutenberg.org/cache/epub/22925/pg22925.txt"
url_list[96] = "http://www.gutenberg.org/files/23395/23395-0.txt"
url_list[97] = "http://www.gutenberg.org/cache/epub/24030/pg24030.txt"
url_list[98] = "http://www.gutenberg.org/cache/epub/24125/pg24125.txt"
url_list[99] = "http://www.gutenberg.org/cache/epub/24656/pg24656.txt"
url_list[100] = "http://www.gutenberg.org/cache/epub/20557/pg20557.txt"

Download all 100 books and place in a list. Then print out their titles.

In [8]:
book_list = list(range(0, max_range)) 
print ("Book Title List:")
for count in range(max_range):
    book_list[count] = text_fetcher(url_list[count])
    print book_list[count][:70]

Book Title List:
The Project Gutenberg EBook of A Primary Reader, by E. Louise Smythe
The Project Gutenberg EBook of The Bird-Woman of the Lewis and Clark
The Project Gutenberg EBook of Dr. Scudder's Tales for Little Readers
The Project Gutenberg EBook of The Louisa Alcott Reader, by Louisa M.
The Project Gutenberg EBook of Boy Blue and His Friends
by Etta Aust
The Project Gutenberg EBook of The Book of Nature Myths, by Florence 
The Project Gutenberg EBook of The Flag of My Country. Shikéyah Bidah
Project Gutenberg's Chambers's Elementary Science Readers, by Various
The Project Gutenberg EBook of The Little Lame Prince, by 
Dinah Mar
The Project Gutenberg EBook of Harry's Ladder to Learning, by Anonymo
The Project Gutenberg EBook of Little Present, by Unknown

This eBo
The Project Gutenberg EBook of The Beacon Second Reader, by James H. 
The Project Gutenberg EBook of The Child's World
by Hetty Browne, Sa
The Project Gutenberg EBook of De La Salle Fifth Reader
by Brother

Tokenize downloaded books and place in a different list for further analysis. Print out first 10 tokens of each book.

In [9]:
tokens_list = list(range(0, max_range)) 
print ("Book Token List:")
for count in range(max_range):
    tokens_list[count] = word_tokenize(book_list[count])
    print tokens_list[count][:10]

Book Token List:
[u'\ufeffThe', u'Project', u'Gutenberg', u'EBook', u'of', u'A', u'Primary', u'Reader', u',', u'by']
[u'\ufeffThe', u'Project', u'Gutenberg', u'EBook', u'of', u'The', u'Bird-Woman', u'of', u'the', u'Lewis']
[u'\ufeffThe', u'Project', u'Gutenberg', u'EBook', u'of', u'Dr.', u'Scudder', u"'s", u'Tales', u'for']
[u'\ufeffThe', u'Project', u'Gutenberg', u'EBook', u'of', u'The', u'Louisa', u'Alcott', u'Reader', u',']
[u'\ufeffThe', u'Project', u'Gutenberg', u'EBook', u'of', u'Boy', u'Blue', u'and', u'His', u'Friends']
[u'\ufeffThe', u'Project', u'Gutenberg', u'EBook', u'of', u'The', u'Book', u'of', u'Nature', u'Myths']
[u'\ufeffThe', u'Project', u'Gutenberg', u'EBook', u'of', u'The', u'Flag', u'of', u'My', u'Country']
[u'\ufeffProject', u'Gutenberg', u"'s", u'Chambers', u"'s", u'Elementary', u'Science', u'Readers', u',', u'by']
[u'\ufeffThe', u'Project', u'Gutenberg', u'EBook', u'of', u'The', u'Little', u'Lame', u'Prince', u',']
[u'\ufeffThe', u'Project', u'Gutenberg', u'EBoo

[u'\ufeffThe', u'Project', u'Gutenberg', u'EBook', u'of', u'Modern', u'Prose', u'And', u'Poetry', u';']
[u'\ufeffThe', u'Project', u'Gutenberg', u'EBook', u'of', u'A', u'Catechism', u'of', u'Familiar', u'Things']
[u'\ufeffThe', u'Project', u'Gutenberg', u'EBook', u'of', u'Children', u"'s", u'Classics', u'in', u'Dramatic']
[u'\ufeffThe', u'Project', u'Gutenberg', u'EBook', u'of', u'The', u'Story', u'of', u'the', u'Mind']
[u'\ufeffThe', u'Project', u'Gutenberg', u'EBook', u'of', u'The', u'Story', u'of', u'Glass', u',']
[u'\ufeffThe', u'Project', u'Gutenberg', u'eBook', u',', u'The', u'Story', u'of', u'Porcelain', u',']
[u'\ufeffThe', u'Project', u'Gutenberg', u'EBook', u'of', u'The', u'Story', u'of', u'Sugar', u',']
[u'\ufeffThe', u'Project', u'Gutenberg', u'EBook', u'of', u'The', u'Story', u'of', u'Wool', u',']
[u'\ufeffProject', u'Gutenberg', u"'s", u'Steve', u'and', u'the', u'Steam', u'Engine', u',', u'by']
[u'\ufeffThe', u'Project', u'Gutenberg', u'eBook', u',', u'Carl', u'and', u'th

1) Vocabulary score function

This function is used for scoring the vocabulary size of a text, and normalizing the score from 0 to 1. 
In order to calculate we first scan all available texts present in a list and find the one with the highest vocabulary number and store it to the variable max_length. We then use that variable dividing it with the equivalent number of the text that interests us to get the vocabulary score. By doing this division with the largest vocabulary number we ensure that our results will be normalized between 1 and 0, since we know that from the available books there is no one with a higher vocabulary length than max_length.

In [10]:
from __future__ import division
def vocab_scorer(text):  
    try:
        max_length = 0
        score = 0.0
        for count in range(max_range):
            if count == 1:
                max_length = len(set(tokens_list[count]))
            if len(set(tokens_list[count])) > max_length:
                max_length = len(set(tokens_list[count]))
        print max_length
        score = len(set(text)) / max_length
        return score
    except (TypeError):
        pass

Calculate the vocabulary score for all books in our list and print out the results.

In [11]:
i = 0
print ("Vocabulary Size Score:")
for tl in tokens_list:
    print ("Book_%d - %f" %  (i, vocab_scorer(tl)))
    i = i + 1 

Vocabulary Size Score:
Book_0 - 0.114671
Book_1 - 0.126029
Book_2 - 0.302526
Book_3 - 0.299281
Book_4 - 0.123131
Book_5 - 0.176150
Book_6 - 0.077761
Book_7 - 0.211670
Book_8 - 0.189014
Book_9 - 0.183393
Book_10 - 0.059045
Book_11 - 0.263240
Book_12 - 0.301020
Book_13 - 0.533086
Book_14 - 0.686464
Book_15 - 0.756982
Book_16 - 0.124464
Book_17 - 0.233805
Book_18 - 0.273207
Book_19 - 0.599954
Book_20 - 0.825530
Book_21 - 1.000000
Book_22 - 0.127361
Book_23 - 0.607255
Book_24 - 0.572372
Book_25 - 0.699733
Book_26 - 0.991077
Book_27 - 0.691563
Book_28 - 0.341928
Book_29 - 0.991946
Book_30 - 0.223143
Book_31 - 0.705064
Book_32 - 0.735717
Book_33 - 0.100649
Book_34 - 0.702225
Book_35 - 0.419689
Book_36 - 0.992873
Book_37 - 0.344941
Book_38 - 0.077761
Book_39 - 0.241164
Book_40 - 0.686812
Book_41 - 0.224765
Book_42 - 0.149322
Book_43 - 0.659926
Book_44 - 0.072720
Book_45 - 0.277668
Book_46 - 0.061653
Book_47 - 0.458744
Book_48 - 0.419747
Book_49 - 0.393035
Book_50 - 0.288678
Book_51 - 0.319156

2) Long-word score function

This function is used for scoring the long-word vocabulary size of a text and normalizing the output from 0 to 1. 
We begin by grabbing a list of all unique words in a text and feed to this function. After that we keep the words that contain more than 6 letters (this is what we decided a long word to be) and then count them. Finally we divide the counted long-words with the total unique words of the text provided and thus calculate the long-word score. Because we divided the counted long-words with the total unique words of the book our result is already normalized between one and zero.

In [12]:
def long_word_scorer(text):
    try:
        sort_words = sorted(set(text))
        count_long_words = 0
        long_word_score = 0.0
        for s in sort_words:
            if len(s) > 6:
                count_long_words = count_long_words + 1
        long_word_score = count_long_words / len(set(text))
        return long_word_score
    except (TypeError):
        pass

Calculate the long-word score for all books in our list and print out the results.

In [13]:
i = 0
print ("Long-word Vocabulary Score:")
for tl in tokens_list:
        print ("Book_%d - %f" %  (i, long_word_scorer(tl)))
        i = i + 1 

Long-word Vocabulary Score:
Book_0 - 0.298130
Book_1 - 0.301609
Book_2 - 0.490136
Book_3 - 0.408906
Book_4 - 0.326118
Book_5 - 0.357566
Book_6 - 0.459762
Book_7 - 0.359431
Book_8 - 0.413857
Book_9 - 0.326698
Book_10 - 0.404318
Book_11 - 0.245212
Book_12 - 0.347449
Book_13 - 0.446957
Book_14 - 0.517599
Book_15 - 0.517912
Book_16 - 0.309125
Book_17 - 0.356877
Book_18 - 0.352492
Book_19 - 0.470639
Book_20 - 0.537236
Book_21 - 0.556090
Book_22 - 0.318016
Book_23 - 0.485019
Book_24 - 0.478842
Book_25 - 0.526664
Book_26 - 0.558115
Book_27 - 0.491663
Book_28 - 0.475513
Book_29 - 0.488405
Book_30 - 0.352636
Book_31 - 0.451759
Book_32 - 0.440104
Book_33 - 0.367876
Book_34 - 0.570509
Book_35 - 0.526992
Book_36 - 0.658302
Book_37 - 0.563413
Book_38 - 0.459762
Book_39 - 0.524748
Book_40 - 0.575719
Book_41 - 0.399330
Book_42 - 0.394645
Book_43 - 0.552639
Book_44 - 0.411952
Book_45 - 0.492905
Book_46 - 0.447368
Book_47 - 0.508400
Book_48 - 0.491579
Book_49 - 0.424149
Book_50 - 0.466479
Book_51 - 0.4

Lexical diversity function

This function was calculated on the previous homework. It is used to calculate the lexical (word) diversity of a text. In order to do so we first calculate the total word count of the provided text and then calculate the unique word count. Finally, the diversity score is generated by dividing the total word count with the unique vocabulary size.

In [14]:
def lexical_diversity(text):
    word_count = len(text)
    vocab_size = len(set(text))
    diversity_score = word_count / vocab_size
    return diversity_score

Normalization function

because the calculated diversity score above is not between 1 and 0, normalization must be applied. This function attempts to normalize any count by dividing it with the max/total number of counts. This generates a number between 0 and 1.

In [15]:
def normalize(count, total):
    return  count / total

3) Text Difficulty function

This function is used for calculating the difficulty of a provided text. In order to do so a combination of all the above functions is used after they have been equally weighted. In other words, the lexical diversity score is initially calculated and then normalized. After that the vocabulary score is calculated and then the long-word score is generated. Finally the text difficulty score is calculated by adding all the above scores and dividing by their count (3). This gives us a normalized final score.

In [16]:
def text_dif_scorer(text):
    try:
        lex_div = lexical_diversity(text)
        nlex_div = normalize(lex_div, len(text)+len(set(text)))
        voc_scr = vocab_scorer(text)
        lwd_scr = long_word_scorer(text)
        txt_dif = nlex_div + voc_scr + lwd_scr / 3
        if txt_dif > 1:
            txt_dif = 1.0
        return txt_dif
    except (TypeError):
        pass

Calculate the text difficulty score for all books in our list and print out the results.

In [18]:
i = 0
print ("Text Difficulty Score:")
for tl in tokens_list:
        print ("Book_%d - %f" %  (i, text_dif_scorer(tl)))
        i = i + 1 

Text Difficulty Score:
Book_0 - 0.214488
Book_1 - 0.226963
Book_2 - 0.466078
Book_3 - 0.435760
Book_4 - 0.232257
Book_5 - 0.295647
Book_6 - 0.231608
Book_7 - 0.331723
Book_8 - 0.327231
Book_9 - 0.292562
Book_10 - 0.194596
Book_11 - 0.345167
Book_12 - 0.417010
Book_13 - 0.682168
Book_14 - 0.859076
Book_15 - 0.929689
Book_16 - 0.227904
Book_17 - 0.352977
Book_18 - 0.390893
Book_19 - 0.756919
Book_20 - 1.004672
Book_21 - 1.185416
Book_22 - 0.233757
Book_23 - 0.769013
Book_24 - 0.732077
Book_25 - 0.875362
Book_26 - 1.177168
Book_27 - 0.855526
Book_28 - 0.500584
Book_29 - 1.154800
Book_30 - 0.340921
Book_31 - 0.855725
Book_32 - 0.882489
Book_33 - 0.223728
Book_34 - 0.892464
Book_35 - 0.595465
Book_36 - 1.212353
Book_37 - 0.532894
Book_38 - 0.231608
Book_39 - 0.416284
Book_40 - 0.878795
Book_41 - 0.358115
Book_42 - 0.281213
Book_43 - 0.844219
Book_44 - 0.210666
Book_45 - 0.442148
Book_46 - 0.211518
Book_47 - 0.628324
Book_48 - 0.583731
Book_49 - 0.534550
Book_50 - 0.444352
Book_51 - 0.465842

Comparing our results with those of last weeks homework, we can see that the scores are vastly different.
Lex Div - HW1  | Txt_Dif - HW2

Texts  Lex_Div  Txt_Dif    Title
Book_12  13      0.417   The Beacon Second Reader
Book_13  20      0.682   The Child's World Third Reader
Book_30  16      0.341   Sanders' Union Fourth Reader


What stands out is the fact that the range of results is vastly different (lexical diversity 0~20 and text difficulty 0~1).
Additionally, another phenomenon that stands out is that Book_12 has a higher text difficulty score than Book_30 even though Book_30 has a higher lexical diversity score. Book 13 has the highest lexical diversity and text difficulty scores as expected.